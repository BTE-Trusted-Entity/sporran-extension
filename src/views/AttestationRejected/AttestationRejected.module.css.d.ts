export const __esModule: true;
export const container: string;
export const heading: string;
export const cardContainer: string;
export const info: string;
export const buttonsLine: string;
export const remove: string;
export const keep: string;
