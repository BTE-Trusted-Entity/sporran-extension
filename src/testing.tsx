export * from '@testing-library/react';
export { waitForElementToBeRemoved } from '@testing-library/dom';
import { act, render as externalRender, screen } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import dialogPolyfill from 'dialog-polyfill';

import { AccountsProviderMock } from './testing/AccountsProviderMock';

export {
  AccountsProviderMock,
  accountsMock,
} from './testing/AccountsProviderMock';

export { mockBackgroundScript } from './testing/mockBackgroundScript';

export function render(
  ui: Parameters<typeof externalRender>[0],
  options?: Parameters<typeof externalRender>[1],
): ReturnType<typeof externalRender> {
  return externalRender(
    <AccountsProviderMock>
      <MemoryRouter>{ui}</MemoryRouter>
    </AccountsProviderMock>,
    options,
  );
}

export async function waitForTooltipUpdate(): Promise<void> {
  expect(await screen.findByRole('status')).toHaveAttribute(
    'data-popper-escaped',
  );
}

const dialogPromise = Promise.resolve();
(dialogPolyfill.registerDialog as jest.Mock).mockReturnValue(dialogPromise);

export async function waitForDialogUpdate(): Promise<void> {
  await act(() => dialogPromise);
}
