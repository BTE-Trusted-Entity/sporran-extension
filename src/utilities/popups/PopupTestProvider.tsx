import { MemoryRouter } from 'react-router-dom';

import { jsonToBase64 } from '../base64/base64';

interface Props {
  path: string;
  data: unknown;
  children: JSX.Element;
}

export function PopupTestProvider({
  path,
  data,
  children,
}: Props): JSX.Element {
  const serialized = jsonToBase64(data);
  return (
    <MemoryRouter initialEntries={[`${path}?data=${serialized}`]}>
      {children}
    </MemoryRouter>
  );
}
