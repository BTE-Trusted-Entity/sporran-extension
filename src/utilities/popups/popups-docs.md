# Integration of websites with the Sporran extension

We need to provide JS api for the websites to be able to interact with the extension.
Apparently the only way for an extension to do that is to inject a script into
the webpage which will create that API.

We also need to show some extension UI in the response to the API calls.
The browsers intentionally prevent JS code from showing the standard extension popup.
Luckily the extension has an internal URI and its UI can be shown in a normal browser
window (or popup window). However only the code running in the extension background
process or popup process can do this. The browsers intentionally prevent JS code
on websites, in injected scripts, as well as in the content scripts from accessing
the extension UI URIs.

Finally, we also need to pass the API call data to the extension UI.
The UI uses internally an in-memory router which is unaware about the actual URI.

This is how the system has to work in such environment:

1. The extension API is exposed to the website JS by an injected script.
   In response to the API calls this script passes messages to the content script.
1. The content script injects the injected script into the webpage.
   It also listens for the messages from it and passes them on to the background process.
1. The background process listens to the messages from the content script.
   When a message is received, it constructs an internal extension URI and
   encodes in it the parameters of the call. Then it opens this URI
   in a new browser window.
1. When the extension UI initializes it detects that it is shown in the browser window,
   not in the extension popup. In that case it takes the parameters encoded in the URI,
   and transforms them for use in the internal router.
