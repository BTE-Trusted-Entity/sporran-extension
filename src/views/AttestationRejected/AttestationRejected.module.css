.container {
  composes: container from '../../components/View/View.module.css';
}

.heading {
  composes: screenHeadline from '../../components/Typography/Typography.module.css';
}

.cardContainer {
  composes: fullWidthContainer from '../../components/View/View.module.css';
  list-style: none;
  max-height: calc(330rem / 16);
}

.info {
  composes: subline from '../../components/Typography/Typography.module.css';
  margin: 1rem 0;
  white-space: pre-line;
}

.buttonsLine {
  composes: buttonsLine from '../../components/Button/Button.module.css';
}

.remove {
  composes: buttonPrimary from '../../components/Button/Button.module.css';
}

.keep {
  composes: buttonSecondary from '../../components/Button/Button.module.css';
}
