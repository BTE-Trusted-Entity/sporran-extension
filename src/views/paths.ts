export { generatePath } from 'react-router-dom';

export const paths = {
  home: '/',
  settings: '/settings',
  access: '/access',
  identity: {
    base: '/identity/',
    add: '/identity/add',
    create: {
      start: '/identity/create',
      backup: '/identity/create/backup',
      verify: '/identity/create/verify',
      password: '/identity/create/password',
    },
    import: {
      start: '/identity/import',
      password: '/identity/import/password',
    },

    overview: '/identity/:address/:type(created|imported|pwreset)?',
    reset: {
      start: '/identity/:address/reset',
      password: '/identity/:address/reset/password',
    },
    send: {
      start: '/identity/:address/send',
      review: '/identity/:address/send/review',
      warning: '/identity/:address/send/warning',
    },
    receive: '/identity/:address/receive',
    credentials: {
      base: '/identity/:address/credentials',
      presentation: '/identity/:address/credentials/:hash/presentation',
    },
    remove: '/identity/:address/remove',
    did: {
      upgrade: {
        start: '/identity/:address/did/upgrade',
        kilt: '/identity/:address/did/upgrade/kilt',
        euro: '/identity/:address/did/upgrade/euro',
      },
      manage: {
        start: '/identity/:address/did',
        downgrade: {
          base: '/identity/:address/did/downgrade',
          warning: {
            web3name: '/identity/:address/did/downgrade/warning/web3name',
            credentials: '/identity/:address/did/downgrade/warning/credentials',
          },
          sign: '/identity/:address/did/downgrade/sign',
        },
        endpoints: {
          start: '/identity/:address/did/endpoints',
          edit: '/identity/:address/did/endpoints/:id?',
          sign: '/identity/:address/did/endpoints/sign',
        },
      },
      repair: '/identity/:address/did/repair',
    },
    web3name: {
      create: {
        base: '/identity/:address/web3name/create',
        info: '/identity/:address/web3name/create/info',
        form: '/identity/:address/web3name/create/form',
        choose: '/identity/:address/web3name/create/choose/:web3name',
        kilt: '/identity/:address/web3name/create/kilt/:web3name',
        euro: '/identity/:address/web3name/create/euro/:web3name',
      },
      manage: {
        start: '/identity/:address/web3name',
        remove: '/identity/:address/web3name/remove',
      },
    },
  },
  popup: {
    base: '/popup',
    import: '/popup/import',
    access: '/popup/access',
    claim: '/identity/:address/claim',
    save: '/popup/save',
    share: {
      start: '/popup/share',
      sign: '/popup/share/sign',
    },
    sign: '/popup/sign',
    signRaw: '/popup/signRaw',
    signDid: {
      start: '/identity/:address/did/sign',
      sign: '/identity/:address/did/sign/password',
      credentials: '/identity/:address/did/sign/credentials',
    },
    signDidExtrinsic: '/identity/:address/did/signExtrinsic',
    createDid: '/identity/:address/did/create',
  },
};
