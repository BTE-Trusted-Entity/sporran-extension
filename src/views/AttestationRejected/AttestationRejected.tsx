import { useCallback, useEffect } from 'react';
import browser from 'webextension-polyfill';
import { find } from 'lodash-es';

import * as styles from './AttestationRejected.module.css';

import { usePopupData } from '../../utilities/popups/usePopupData';
import {
  deleteCredential,
  saveCredential,
  useCredentials,
} from '../../utilities/credentials/credentials';
import { rejectChannel } from '../../channels/rejectChannel/rejectChannel';
import { CredentialCard } from '../../components/CredentialCard/CredentialCard';
import { UnknownCredential } from '../../components/UnknownCredential/UnknownCredential';
import { RejectInput } from '../../channels/rejectChannel/types';

export function AttestationRejected() {
  const t = browser.i18n.getMessage;

  const rootHash = usePopupData<RejectInput>();

  const sporranCredentials = useCredentials();
  const sporranCredential = find(sporranCredentials, {
    credential: { rootHash },
  });

  const handleRemoveClick = useCallback(async () => {
    if (sporranCredential) {
      await deleteCredential(sporranCredential);
    }
    window.close();
  }, [sporranCredential]);

  const handleKeepClick = useCallback(() => {
    window.close();
  }, []);

  useEffect(() => {
    (async () => {
      await rejectChannel.return(undefined);
      if (sporranCredential) {
        await saveCredential({ ...sporranCredential, status: 'rejected' });
      }
    })();
  }, [sporranCredential]);

  if (!sporranCredentials) {
    return null; // storage data pending
  }

  if (!sporranCredential) {
    return <UnknownCredential rootHash={rootHash} />;
  }

  return (
    <main className={styles.container}>
      <h1 className={styles.heading}>
        {t('view_AttestationRejected_heading')}
      </h1>

      <p className={styles.info}>{t('view_AttestationRejected_info')}</p>

      <section className={styles.cardContainer}>
        <CredentialCard
          sporranCredential={sporranCredential}
          expand
          collapsible={false}
          buttons={false}
        />
      </section>

      <p className={styles.buttonsLine}>
        <button type="button" className={styles.keep} onClick={handleKeepClick}>
          {t('view_AttestationRejected_keep')}
        </button>
        <button
          type="button"
          className={styles.remove}
          onClick={handleRemoveClick}
        >
          {t('view_AttestationRejected_remove')}
        </button>
      </p>
    </main>
  );
}
