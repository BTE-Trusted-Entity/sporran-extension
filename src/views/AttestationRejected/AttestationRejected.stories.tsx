import { Meta } from '@storybook/react';

import { PopupTestProvider } from '../../utilities/popups/PopupTestProvider';
import { credentialsMock } from '../../utilities/credentials/CredentialsProvider.mock';
import { paths } from '../paths';

import { AttestationRejected } from './AttestationRejected';

export default {
  title: 'Views/AttestationRejected',
  component: AttestationRejected,
} as Meta;

export function Template(): JSX.Element {
  return (
    <PopupTestProvider
      path={paths.popup.reject}
      data={credentialsMock[13].credential.rootHash}
    >
      <AttestationRejected />
    </PopupTestProvider>
  );
}
