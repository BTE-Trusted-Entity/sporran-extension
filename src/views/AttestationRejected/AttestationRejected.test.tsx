import { render } from '../../testing/testing';

import { paths } from '../paths';
import { credentialsMock } from '../../utilities/credentials/CredentialsProvider.mock';
import { PopupTestProvider } from '../../utilities/popups/PopupTestProvider';

import { AttestationRejected } from './AttestationRejected';

describe('AttestationRejected', () => {
  it('should match the snapshot', async () => {
    const { container } = render(
      <PopupTestProvider
        path={paths.popup.reject}
        data={credentialsMock[13].credential.rootHash}
      >
        <AttestationRejected />
      </PopupTestProvider>,
    );
    expect(container).toMatchSnapshot();
  });
});
