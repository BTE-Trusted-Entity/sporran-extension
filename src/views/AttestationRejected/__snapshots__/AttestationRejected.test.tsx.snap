// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AttestationRejected should match the snapshot 1`] = `
<div>
  <main
    class="container"
  >
    <h1
      class="heading"
    >
      Attestation rejected
    </h1>
    <p
      class="info"
    >
      The attestation you requested was rejected.
The credential will likely not be useful in the future.
Do you want to remove it from your Sporran?
    </p>
    <section
      class="cardContainer"
    >
      <li
        aria-expanded="true"
        class="unusableCredential"
      >
        <section
          class="expanded"
        >
          <section
            class="buttons"
          />
          <dl
            class="details"
          >
            <div
              class="detail"
            >
              <dt
                class="detailName"
              >
                Credential name
              </dt>
              <dd
                class="nameValue"
              >
                GitHub 2
                <button
                  aria-label="Edit credential name"
                  class="editName"
                />
              </dd>
            </div>
            <div
              class="detail"
            >
              <dt
                class="detailName"
              >
                Status
              </dt>
              <dd
                class="detailValue unusableValue"
              >
                Rejected
              </dd>
            </div>
            <div
              class="detail"
            >
              <dt
                class="detailName"
              >
                Username
              </dt>
              <dd
                class="detailValue"
              >
                arty-name
              </dd>
            </div>
            <div
              class="detail"
            >
              <dt
                class="detailName"
              >
                User ID
              </dt>
              <dd
                class="detailValue"
              >
                133055
              </dd>
            </div>
          </dl>
          <h4
            class="technical"
          >
            Technical details
          </h4>
          <dl
            class="details"
          >
            <div
              class="detail"
            >
              <dt
                class="detailName"
              >
                CType
              </dt>
              <dd
                class="detailValue"
              >
                GitHub
              </dd>
            </div>
            <div
              class="detail"
            >
              <dt
                class="detailName"
              >
                Attester
              </dt>
              <dd
                class="detailValue"
              >
                SocialKYC
              </dd>
            </div>
            <div
              class="hash"
            >
              <dt
                class="detailName"
              >
                Credential hash
              </dt>
              <dd
                class="detailValue"
              >
                0x202b70def75caa7d2130524b12d759e711ebf75960e838cbbc27d657560e6675
              </dd>
            </div>
          </dl>
        </section>
        <div />
      </li>
    </section>
    <p
      class="buttonsLine"
    >
      <button
        class="keep"
        type="button"
      >
        Keep
      </button>
      <button
        class="remove"
        type="button"
      >
        Remove
      </button>
    </p>
  </main>
</div>
`;
